numpy/_core/src/umath/umathmodule.c

```diff
@@ -144,15 +145,75 @@ ufunc_frompyfunc(PyObject *NPY_UNUSED(dummy), PyObject *args, PyObject *kwds) {
     /* Do a better job someday */
     doc = "dynamic ufunc based on a python function";
 
+    /* Create the ufunc but without any loops by not passing types*/
     self = (PyUFuncObject *)PyUFunc_FromFuncAndDataAndSignatureAndIdentity(
             (PyUFuncGenericFunction *)pyfunc_functions, data,
-            types, /* ntypes */ 1, nin, nout, identity ? PyUFunc_IdentityValue : PyUFunc_None,
+            NULL, /* ntypes */ 0, nin, nout, identity ? PyUFunc_IdentityValue : PyUFunc_None,
             str, doc, /* unused */ 0, NULL, identity);
 
     if (self == NULL) {
         PyArray_free(ptr);
         return NULL;
     }
+
+    PyArray_DTypeMeta *op_dtypes[NPY_MAXARGS];
+    for (int arg = 0; arg < nin + nout; arg++) {
+        op_dtypes[arg] = PyArray_DTypeFromTypeNum(types[arg]);
```

Comment:
if you scroll up, you will see that types are all object, so you can delete it there.
(together with things like `pyfunc_functions` and `data`.)

Reply:
Good point.